compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java

```diff
@@ -549,6 +555,72 @@ private boolean isValidReturnBType(Class<?> jType, BType bType, JMethodRequest j
                 this.symbolTable.arrayJsonType};
     }
 
+    private boolean isReadOnlyCompatibleReturnType(Class<?> jType, JMethodRequest jMethodRequest)
+            throws ClassNotFoundException {
+        if (jType.getTypeName().equals(J_OBJECT_TNAME)) {
+            return true;
+        }
+
+        for (BType member : getDefinedReadOnlyMemberTypes()) {
+            if (isValidReturnBType(jType, member, jMethodRequest)) {
+                return true;
+            }
+        }
+
+        if (this.classLoader.loadClass(BError.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BFunctionPointer.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BObject.class.getCanonicalName()).isAssignableFrom(jType)) {
+            // Service and Object.
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BTypedesc.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BHandle.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BXML.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        if (this.isValidListType(jType, jMethodRequest)) {
+            return true;
+        }
+
+        if (this.classLoader.loadClass(BMap.class.getCanonicalName()).isAssignableFrom(jType)) {
+            return true;
+        }
+
+        return this.classLoader.loadClass(TableValue.class.getCanonicalName()).isAssignableFrom(jType);
+    }
+
+    private BType[] getDefinedReadOnlyMemberTypes() {
+        return new BType[]{
```

Comment:
Can this be a constant? Do we have to do this for each and every time?

Reply:
Made this a final variable, initialized in the constructor (since we need the symbol table to retrieve the members).